sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/ReadFromKafkaDoFn.java

```diff
@@ -461,6 +461,11 @@ public ProcessContinuation processElement(
           return ProcessContinuation.resume();
         }
         for (ConsumerRecord<byte[], byte[]> rawRecord : rawRecords) {
+          // If the Kafka consumer returns a record with an offset that is already processed
+          // the record can be safely skipped.
+          if (rawRecord.offset() < startOffset) {
+            continue;
```

Comment:
I'm concerned that if we have some case where seek just stops working in general this will just end up scanning a bunch of records and be hard to read.  Should we count and log how many skipped records due to the seek failing? Or we could handle differently like attempting to seek again or returning ProcessContinuation.resume() to try the seek again later.

Reply:
Yeah, that makes sense. But I am not super familiar with `ProcessContinuation.resumer()`.  From the docs it seems that the DoFn will begin processing the same element again, does that not cause an infinite loop since the element is already processed.

Maybe if the processing begins from the beginning of the DoFn and we seek again then maybe we can get a different result.

Reply:
Yeah, that makes sense. I am not super familiar with how `ProcessContinuation` works, the docs say that it begins processing the DoFn at the same element. Would that not result in an infinite loop, since we have already processed that element.

Or does the processing begin at the beginning of the DoFn, which would mean we seek again and maybe get the records that we need this time? If that is the case, does it make sense to resume with a delay? And also instead of doing this here we can do it immediately after the seek by checking the position.

Reply:
ProcessContinuation.resume() will start again from the same restriction at a later point.  The bundle stops processing and the runner will reschedule the restriction for processing later. Even without a delay it gives a chance for other processing to start, but a delay in this case is probably desirable because we observed it took several minutes to resolve.

So if the seek never works, it would be a problem, but if the seek eventually works (assuming it is some delay in noticing the partition offsets) that would be a way to wait for it without skipping/reading a bunch of events.

+1 to putting the logic to handle the failed seek (either a resume or skipping records) right after the seek itself instead of in the main read loop. Perhaps we should read and skip, counting how many we skipped and log a warning that X records were read/skipped due to seek not returning to the same poisition.  And if we do that for 10 seconds or something and don't reach the start offset we can stop and return a PRocessingContinuation.resume() with a delay.

Reply:
Done. Skip till 10 sec, then stop with backoff of 5 minutes. I chose the values arbitrarily, i can change it if you have a better default suggestion.